### ISSUE TITLE
tap cli runner badly munges source locations

### ISSUE THREAD

**Issue by Fishrock123** [tapjs/tapjs#584]
Encountered when running `tap tests/*.js -J`:

```
 FAIL  tests/test-streams3-zlib.js
 ✖ Exit Callback received unexpected error

  tests/test-streams3-zlib.js
  1 | 'use strict'
> 2 |
  3 | const tap = require('tap')
  4 | const zlib = require('zlib')
  5 |

  test: tests/test-streams3-zlib.js test streams3 (BobDuplex) with a Transform
  found:
    !error
    name: Error
    message: "WritableSource: pull() called without having a write callback"
    stack: >-
      Error: WritableSource: pull() called without having a write callback
          at WritableSource.pull (/Users/Jeremiah/Documents/bob/helpers/writable-source.js:20:90)
          at Immediate.<anonymous> (/Users/Jeremiah/Documents/bob/reference-verify.js:18:683)
          at processImmediate (internal/timers.js:439:21)
          at process.topLevelDomainCallback (domain.js:126:23)
  stack: |
    AssertionSink.exitCb (tests/test-streams3-zlib.js:2:398)
    AssertionSink.next (tests/helpers/assertion-sink.js:4:525)
    Immediate._onImmediate (reference-verify.js:12:498)
```

Those user source locations are completely wrong.

Running as `node tests/test-streams3-zlib.js` gives the correct source locations:
```
# Subtest: test streams3 (BobDuplex) with a Transform
    1..1
    not ok 1 - Exit Callback received unexpected error
      ---
      found:
        !error
        name: Error
        message: "WritableSource: pull() called without having a write callback"
        stack: >-
          Error: WritableSource: pull() called without having a write callback
              at WritableSource.pull (/Users/Jeremiah/Documents/bob/helpers/writable-source.js:149:17)
              at Immediate.<anonymous> (/Users/Jeremiah/Documents/bob/reference-verify.js:184:19)
              at processImmediate (internal/timers.js:439:21)
      at:
        line: 43
        column: 7
        file: tests/test-streams3-zlib.js
        type: AssertionSink
        method: exitCb
      stack: |
        AssertionSink.exitCb (tests/test-streams3-zlib.js:43:7)
        AssertionSink.next (tests/helpers/assertion-sink.js:32:51)
        Immediate._onImmediate (reference-verify.js:137:17)
      source: |2
          wstream.start(error => {
            t.error(error, 'Exit Callback received unexpected error')
        ------^
            t.end()
          })
      ...

    # failed 1 test
```

**Comment by isaacs** [comment]
Willing to bet it's because of coverage.  Do you get the right location when you run it with `--no-cov`?

**Comment by petekneller** [comment]
I'm guessing I've just stumbled into the same problem. And it does look like coverage is the common factor. In my case its Typescript. The output JS is in `dist/` and the sourcemaps are alongside.

Straight from the tap lib: `> node dist/test/obstacles/defaults.js`
```
....
 stack: |
   Object.addDefaults [as default] (src/obstacles/defaults.ts:10:36)
   Object.constructor [as default] (src/obstacles/index.ts:8:21)
   Suite.t.context [as fn] (test/obstacles/defaults.js:11:33)
```

From the CLI: `> npx tap -Rtap dist/test/obstacles/defaults.js`
```
....
 stack: |
   Object.addDefaults [as default] (dist/src/obstacles/defaults.js:1:11203)
   Object.constructor [as default] (dist/src/obstacles/index.js:1:6019)
   Suite.t.context [as fn] (dist/test/obstacles/defaults.js:1:9888)
```

And without coverage: `> npx tap -Rtap --no-coverage  dist/test/obstacles/defaults.js`
```
....
 stack: |
   Object.addDefaults [as default] (src/obstacles/defaults.ts:10:36)
   Object.constructor [as default] (src/obstacles/index.ts:8:21)
   Suite.t.context [as fn] (test/obstacles/defaults.js:11:33)
```

I'm quite new to tap and nyc. @isaacs any thoughts on where to begin?

**Comment by Fishrock123** [comment]
I think this worked without coverage but I cannot recall exactly. Sorry. 😞

**Comment by isaacs** [comment]
Sorry it took a while to cycle back to this.

Can you try with the latest release, 14.7.1?  It _should_ be properly generating and interpreting source maps now, on all supported versions of node.

**Comment by Fishrock123** [comment]
This issue still occurs for me with tap 14.7.1.

Apply this patch:
```patch
From 8192b98968195d0593ba7a04edc82f67ba12b25c Mon Sep 17 00:00:00 2001
From: Jeremiah Senkpiel <fishrock123@rocketmail.com>
Date: Tue, 15 Oct 2019 19:37:11 -0700
Subject: [PATCH] cause problematic stack output in node-tap

---
 reference-verify.js | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/reference-verify.js b/reference-verify.js
index 79e6600..335528c 100644
--- a/reference-verify.js
+++ b/reference-verify.js
@@ -133,6 +133,8 @@ class Verify {
       this[kHadError] = true
     }
 
+    this.sink.next(status_type.error, new Error('OwO'))
+
     setImmediate(_ => {
       this.sink.next(status, error, buffer, bytes)
     })
-- 
2.17.1
```

To this repo: https://github.com/Fishrock123/bob

**Comment by isaacs** [comment]
Ah, I see what's going on.  Two things, actually.

First, your test files are getting covered.  That's always a recipe for weirdness.

However, it wouldn't be an issue given the fixes in 14.7.1, _except_ for the fact that you're using `t.error()` rather than just throwing the error and letting the domain handle it.

`t.error(er)` is a bit weird, because it's an assert.  Typically, you want to see a pointer to the place where the failed assert was.  Eg:

```
$ tap fail.js
 FAIL  fail.js
 ✖ math should work this way

  fail.js
  1 | const t = require('tap')
  2 | t.test('failer', t => {
> 3 |   t.equal(1, 2, 'math should work this way')
    | ----^
  4 |   t.end()
  5 | })

  --- wanted
  +++ found
  -2
  +1
```

But with `t.error()` the error itself _also_ has a stack trace, so maybe it's better to indicate where the error originated?  I could see arguments for either one, but I'm leaning towards making `t.error()` report the same as a throw: show the errors provenance, not the place where you failed an assert on it.

Which of these do you prefer?

```
// provenance style
 FAIL  tests/test-file-read.js
 ✖ Exit Callback received unexpected error

  reference-verify.js
  134 |     }
  135 |
> 136 |     this.sink.next(status_type.error, new Error('OwO'))
      | --------------------------------------^
  137 |
  138 |     setImmediate(_ => {
  139 |       this.sink.next(status, error, buffer, bytes)

  test: tests/test-file-read.js test file read
  stack: |
    Verify.next (reference-verify.js:136:39)
    node_modules/fs-source/index.js:100:21
```

Or, assert indication style:

```
 FAIL  tests/test-file-read.js
 ✖ Exit Callback received unexpected error

  tests/test-file-read.js
  22 |   const stream = new Stream(fileSource, new Verify(), sink)
  23 |   stream.start(error => {
> 24 |     t.error(error, 'Exit Callback received unexpected error')
     | ------^
  25 |   })
  26 | })

  test: tests/test-file-read.js test file read
  found:
    !error
    name: Error
    message: OwO
    stack: |-
      Error: OwO
          at Verify.next (/Users/isaacs/dev/js/bob/reference-verify.js:136:39)
          at /Users/isaacs/dev/js/bob/node_modules/fs-source/index.js:100:21
          at FSReqCallback.wrapper [as oncomplete] (fs.js:470:5)
  stack: |
    AssertionSink.exitCb (tests/test-file-read.js:24:7)
    AssertionSink.exitCb [as next] (tests/helpers/assertion-sink.js:32:51)
    Verify.next (reference-verify.js:136:15)
    node_modules/fs-source/index.js:100:21
```

**Comment by isaacs** [comment]
Yet another approach would be to blend the two.  Show the pointer to the failed assertion, then a "error origin" block underneath it.  Would be more complicated logic to add to the reporter, so I'm not as enthusiastic about it.  Might also be kind of a lot of visual clutter.

**Comment by isaacs** [comment]
Actually it's not that hard, turns out.

Too noisy, you think?

```
 FAIL  tests/test-zlib-transform.js
 ✖ Exit Callback received unexpected error

  tests/test-zlib-transform.js
  35 |   const stream = new Stream(source, new Verify(), zlibTransform, new Verify(), sink)
  36 |   stream.start(error => {
> 37 |     t.error(error, 'Exit Callback received unexpected error')
     | ------^
  38 |     t.end()
  39 |   })
  40 | })

  test: tests/test-zlib-transform.js test zlib transform
  stack: |
    AssertionSink.exitCb (tests/test-zlib-transform.js:37:7)
    AssertionSink.exitCb [as next] (tests/helpers/assertion-sink.js:32:51)
    Verify.next (reference-verify.js:78:17)
    ZlibTransform.next (node_modules/zlib-transform/index.js:228:24)
    Verify.next (reference-verify.js:78:17)
    AssertionSource.next (tests/helpers/assertion-source.js:30:24)
    Immediate.pull [as _onImmediate] (reference-verify.js:186:19)

  Error Origin:

  reference-verify.js
  101 |       // buffer
  102 |       if (error !== null && !Buffer.isBuffer(buffer)) {
> 103 |         throw new TypeError(`[verify] buffer passed to next() was not a Buffer: ${buffer}`)
      | --------------^
  104 |       }
  105 |       // bytes
  106 |       if (error !== null && typeof bytes !== 'number') {

  stack: |
    Verify.next (reference-verify.js:103:15)
    ZlibTransform.next (node_modules/zlib-transform/index.js:228:24)
    Verify.next (reference-verify.js:136:15)
    AssertionSource.next (helpers/buffer-source.js:59:17)
    AssertionSource.pull (tests/helpers/assertion-source.js:56:11)
    Immediate.pull (reference-verify.js:186:19)
  type: TypeError
  message: "[verify] buffer passed to next() was not a Buffer: undefined"
```

**Comment by Fishrock123** [comment]
> First, your test files are getting covered. That's always a recipe for weirdness.

Any idea why? I am only invoking `tap tests/*.js -J`..., nothing weird in my `ENV` either.

> except for the fact that you're using `t.error()` rather than just throwing the error and letting the domain handle it.

Hmm, I could do that, but I'd rather be explicit. I know where the error will be passed to and I was hoping to be able to pass that to `tap`...

Edit: I guess that's off topic-ish. Not quite sure what I prefer for output yet.

**Comment by isaacs** [comment]
Nyc automatically excludes `test/*`, but not `tests/`. Next version of tap will tell nyc to exclude all its test files. 

> Edit: I guess that's off topic-ish. Not quite sure what I prefer for output yet.

No worries :)

I'm leaning towards the double-output style. If you just wanna print the assert or just the origin, there are other options.

**Comment by isaacs** [comment]
Ok!  This constellation of issues should all be fixed on 14.7.2.  Let me know if you still have issues in this area.  Thanks!
